### ISSUE TITLE
Dart2js deferred loading support

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/build#814]
Eventually we need to figure out how to support dart2js with deferred loading. If at all possible this should not involve anything hardcoded the `build` or `build_runner` packages, and could be completely isolated to the `build_compilers` package.

Some ideas and previous solutions include:

- Output archive files and add logic for serving from archive files with specific uri patterns. We would also need specific logic for expanding these archives when creating merged output directories.
- Require users to hardcode the number of deferred outputs they expect. We would need to give good errors when this number needs to be updated. We could allow users to over-declare this number though.
  - This would still be blocked on user config being able to override the output extensions of builders, ideally on a per-target basis.
- Hardcode a certain number of potential output files, and just omit most of them. This doesn't scale though, and also involves some amount of cpu/memory overhead.
- Add some way of a builder declaring output patterns as opposed to specific outputs. This would mean the asset graph would have to become a lot more flexible (today it is completely fixed before a build starts). This also makes checking for overlapping outputs more expensive and complicated.

**Comment by natebosch** [comment]
I like archive files.

I think we've used both archives and hardcoding number of deferred parts in the config with bazel right? IIRC archives are our current preferred approach in bazel and so if we can support that here we'll drive alignment.

Archives are also a more generally useful solution if there are other builders that want this flexibility in outputs (which can't be read during the rest of the build)

**Comment by jakemac53** [comment]
The main issue with archive files is figuring out how and what to serve from them.

We would need to add some sort of special case logic, maybe a special type of asset node, which contains information about what files can actually be served by the tar file. We could potentially create synthetic asset nodes as well which just point at a tar file for their contents, and can't be used as inputs to other actions.

**Comment by natebosch** [comment]
Would it work to force a certain layout for the archive and serve any file in the archive which isn't available as a normal asset?

Request comes in for a path, translate to assetId and check if it's in the graph. If it isn't translate to an archive path and walk the archives we have 1 by 1 and serve the file from the first archive that has it. Archives can be walked in buildaction order.

**Comment by jakemac53** [comment]
We really don't want to be searching all potential archives _on each request_ like that though. We need to know ahead of time what archive to serve from.

For instance you would potentially have one of these archives _per test_, which could be very slow.

**Comment by natebosch** [comment]
We can read the list of files ahead of time and cache that. We'd only read the list of files once per build, not each request

**Comment by natebosch** [comment]
And we can even choose to read only the list of files from the archives that actually changed. I think we can avoid too much performance impact

**Comment by jakemac53** [comment]
If we are reading in the files though, then we might as well create synthetic asset nodes which know exactly where to look?

**Comment by jodinathan** [comment]
does this mean we can't use deferred components with and without @deferred for now?

**Comment by jakemac53** [comment]
Yes, as of now the dart2js support does not work with any form of deferred loading

**Comment by jodinathan** [comment]
this is very important to me. I can not even test my projects without this.
(I could, but I would have to change a lot of places only for this)

**Comment by jakemac53** [comment]
You should be able to test your projects using the dartdevc compiler (recommended for most tests as it is actually _more_ strict) - but I understand that running at least some integration tests with dart2js is a good idea (and you would need it to deploy in any case).

As a temporary workaround you can use the `--output=<some-dir>` option to get a full output directory, and then run dart2js manually within there. You would need to either delete the `.dart.js` files that dartdevc creates or remove your dev dependency on `build_web_compilers` to stop js compilation, but it should at least give you some path for deployment with dart2js in the meantime.

We can try and move fairly on this after dartconf, it is definitely a high priority issue.
